tests/bluetooth/host/keys/bt_keys_add_type/src/main.c

```diff
@@ -0,0 +1,122 @@
+
+	{ BT_ADDR_ID_1, BT_ADDR_LE_1, BT_KEYS_PERIPH_LTK },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_1, BT_KEYS_PERIPH_LTK },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_2, BT_KEYS_IRK },
+	{ BT_ADDR_ID_1, BT_ADDR_LE_3, BT_KEYS_IRK },
+
+	{ BT_ADDR_ID_2, BT_ADDR_LE_1, BT_KEYS_LTK },
+	{ BT_ADDR_ID_2, BT_RPA_ADDR_LE_2, BT_KEYS_IRK },
+	{ BT_ADDR_ID_2, BT_RPA_ADDR_LE_3, BT_KEYS_IRK },
+	{ BT_ADDR_ID_2, BT_ADDR_LE_2, BT_KEYS_LOCAL_CSRK },
+
+	{ BT_ADDR_ID_3, BT_ADDR_LE_1, BT_KEYS_REMOTE_CSRK },
+	{ BT_ADDR_ID_3, BT_ADDR_LE_2, BT_KEYS_LTK_P256 },
+
+	{ BT_ADDR_ID_4, BT_ADDR_LE_1, BT_KEYS_ALL }
+};
+
+/* Global iterator to iterate over the ID, Address and type LUT */
+static uint8_t params_it;
+
+/* Pointer to the current set of testing parameters */
+struct id_addr_type const *current_params_vector;
+
+/* This list will hold returned references while filling keys pool */
+static struct bt_keys *returned_keys_refs[CONFIG_BT_MAX_PAIRED];
+
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_type_lut) == CONFIG_BT_MAX_PAIRED);
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_type_lut) == ARRAY_SIZE(returned_keys_refs));
+
+/*
+ *  Set key type and verify the value
+ *
+ *  Constraints:
+ *   - Full keys pool list
+ *
+ *  Expected behaviour:
+ *   - The key type value is set correctly
+ */
+void test_set_type_value_correctly(void)
+{
+	struct bt_keys *key_ref;
+	int type = current_params_vector->type;
+
+	/* 'params_it' is incremented before running the test case */
+	key_ref = returned_keys_refs[params_it - 1];
+
+	bt_keys_add_type(key_ref, type);
+
+	zassert_true(key_ref->keys == type, "bt_keys_add_type() set incorrect key type value");
+}
+
+/* Initialize test variables */
+static void reset_global_test_round_params(void)
+{
+	params_it = 0;
+	current_params_vector = NULL;
+}
+
+/* Fill the keys pool with testing (ID, Address) pairs */
+static void keys_fill_pool(void)
+{
+	uint8_t id;
+	bt_addr_le_t *addr;
+	struct id_addr_type const *params_vector;
+
+	zassert_true(check_key_pool_is_empty(),
+		     "List isn't empty, make sure to run this test just after a fresh start");
+
+	for (size_t it = 0; it < ARRAY_SIZE(testing_id_addr_type_lut); it++) {
+		params_vector = &testing_id_addr_type_lut[it];
+		id = params_vector->id;
+		addr = params_vector->addr;
+		returned_keys_refs[it] = bt_keys_get_addr(id, addr);
+		zassert_true(returned_keys_refs[it] != NULL,
+			     "bt_keys_get_addr() returned a non-valid reference");
+	}
+}
+
+/* Setup test variables */
+static void unit_test_setup(void)
+{
+	zassert_true((params_it < (ARRAY_SIZE(testing_id_addr_type_lut))),
+		     "Invalid testing parameters index %u", params_it);
+	current_params_vector = &testing_id_addr_type_lut[params_it];
+	params_it++;
+}
+
+void test_main(void)
+{
+
+	reset_global_test_round_params();
+	clear_key_pool();
+	keys_fill_pool();
+
+	ztest_test_suite(bt_keys_add_type_verify_type_value,
+			 LISTIFY(CONFIG_BT_MAX_PAIRED, REGISTER_SETUP_TEARDOWN, (,),
```

Comment:
I don't understand the value of this parameterization over set of max pairings. Firstly, the set of possible combinations to test could easily be larger or smaller than CONFIG_BT_MAX_PAIRED, secondly any attempt to re-order the tests will fail because this is simply repeaiting the same configuration CONFIG_BT_MAX_PAIRED times against params_it, which is not under ztest's control. It should be equivalent and simpler to have one test that simply iterates the scenarios defined by testing_id_addr_type_lut. I suppose you would have any assert terminate before all tests run, which could require multiple iterations to identify whether the tests passes (after a change which breaks multiple behaviors), but that seems a design limitation of ztest (no EXPECT_xxx as in gtest or other soft-assert mechanism).

Reply:
This is an old debate about what is supported by ZTest?

The problem was "How can I run the same test case with a set of input parameters?"

There are some limitations by ZTest

1. Test case function can't have any arguments and it must have the form `void test_name(void)`
2. You can run the UT project with different parameters by adding multiple records to the `testcase.yaml`.
    But, it doesn't support to define a list of parameters within a certain record.
    You will end up with a lot of records equals to the number of parameters you need to run

- One way, as you mentioned, was to run a single test case and iterate over the items using a loop.
  The problem was that the output report will include a single record for that test case with the result pass or fail for all 
  parameters, so you won't be able to recognize which iterations failed.

- The second approach, which is currently used, was to duplicate the test case and depend on a global iterator to select the parameters to be tested.

Althoug I was in favor of the second approach by that time and they agreed on it, I'm not any more.
It's not easy to follow and harder to re-order.

Actually, now I'm looking at the use of new ZTest APIs and I'm trying to look for another way to do it away from duplicating test cases.

Reply:
> I don't understand the value of this parameterization over set of max pairings.

`CONFIG_BT_MAX_PAIRED` is just used to set number of iterations

Reply:
Except that CONFIG_BT_MAX_PAIRED is 16 (I think), but could probably be set as little as 2. The combinations aren't important to the current implementation of bt_keys_add_type, but if they were, you could have lots more combinations. Given a static and random address, LTK or 256-bit LTK, presence or absence of local IRK, presence or absence of local CSRK, presence or absence of remote IRK, presence or absence of remote CSRK... that's 64 combinations already.

Reply:
I already changed that based on your previous comment.
Now I'm iterating of the list in one single test case and added a new one to test the value of `BT_KEYS_ALL`